src/types/widget.ts

```diff
@@ -328,7 +284,73 @@ export type Widget = {
     state: SearchParameters,
     widgetSearchParametersOptions: WidgetSearchParametersOptions
   ): SearchParameters;
-};
+} & (
+  | {
+      $$type: 'ais.searchBox';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => SearchBoxWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.autocomplete';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => AutocompleteWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.breadcrumb';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => BreadcrumbWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.clearRefinements';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => ClearRefinementsWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.configure';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => ConfigureWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.currentRefinements';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => CurrentRefinementsWidgetRenderState;
+    }
+  | {
+      $$type: 'ais.hierarchicalMenu';
+      getWidgetRenderState?: (
+        renderOptions: InitOptions | RenderOptions
+      ) => HierarchicalMenuWidgetRenderState;
+    }
```

Comment:
what if this part is a generic that the connector passes to widget? it can have a default of `{}` or something like that?

Reply:
I think that's possible too. Then the definition of the connector will become something like:

```js
// from
export type AutocompleteConnector = Connector<
  AutocompleteRendererOptions,
  AutocompleteConnectorParams
>;

// to
type AutocompleteXYZ = {
  $$type: 'ais.autocomplete';
  getWidgetRenderState?: (
    renderOptions: InitOptions | RenderOptions
  ) => AutocompleteWidgetRenderState;
};

export type AutocompleteConnector = Connector<
  AutocompleteRendererOptions,
  AutocompleteConnectorParams,
  AutocompleteXYZ,
>;

// with

export type WidgetFactory<TConnectorParams, TWidgetParams, TXyz> = (
  /**
   * The params of the widget.
   */
  widgetParams: TConnectorParams & TWidgetParams
) => Widget & TXyz;
```